File: src/main/java/xyz/emirdev/echogen/utils/ClassUtils.java
package xyz.emirdev.echogen.utils;

import io.github.classgraph.ClassGraph;
import io.github.classgraph.ClassInfo;
import io.github.classgraph.ScanResult;
import xyz.emirdev.echogen.Echogen;

import java.util.function.Consumer;
import java.util.function.Predicate;

public class ClassUtils {

    public static void findClasses(String pkg, Predicate<ClassInfo> condition, Consumer<ClassInfo> function) {
        try (ScanResult scanResult = new ClassGraph()
                .acceptPackages(pkg)
                .addClassLoader(Echogen.get().getClass().getClassLoader())
                .enableClassInfo()
                .scan()) {
            scanResult.getAllClasses().forEach(classinfo -> {
                if (!condition.test(classinfo))
                    return;

                function.accept(classinfo);
            });
        }

    }
}

File: src/main/kotlin/cat/emir/echogen/utils/ClassUtils.kt
package cat.emir.echogen.utils

import cat.emir.echogen.Echogen
import io.github.classgraph.ClassGraph
import io.github.classgraph.ClassInfo

class ClassUtils {
    companion object {
        fun findClasses(pkg: String, condition: (ClassInfo) -> Boolean, function: (ClassInfo) -> Unit) {
            ClassGraph()
                .acceptPackages(pkg)
                .addClassLoader(Echogen.instance.javaClass.classLoader)
                .enableClassInfo()
                .scan().use { scanResult ->
                    scanResult.allClasses.forEach {
                        if (!condition(it)) return@forEach
                        function(it)
                    }
                }
        }
    }
}